# K-Means Clustering

In [62]:
import numpy as np
import random as rd
import math
import matplotlib.pyplot as plt

In [63]:
x = [[5.9, 3.2], [4.6, 2.9], [6.2, 2.8], [4.7, 3.2], [5.5, 4.2], [5.0, 3.0], [4.9, 3.1], [6.7, 3.1], [5.1, 3.8], [6.0, 3.0]]

In [64]:
def euclidean_distance(p1, p2):
    dist = math.sqrt((p1[0]-p2[0])**2+(p1[1]-p2[1])**2)
    return dist

def minimum(one, two, three):
    if one < two and one < three:
        return 1
    if two < one and two < three:
        return 2
    if three < one and three < two:
        return 3
    else:
        return 4

In [65]:
centroid = [[6.2, 3.2], [6.6, 3.7], [6.5, 3.0]]
global cluster1 # red
global cluster2 # green
global cluster3 # blue

cluster1= [] # red
cluster2= [] # green
cluster3= [] # blue

def make_clusters(arr):
    for each in arr:
        one = euclidean_distance(centroid[0], each)
        two = euclidean_distance(centroid[1], each)
        three = euclidean_distance(centroid[2], each)
        what = minimum(one, two, three)
        if what == 1:
            cluster1.append(each)
        if what == 2:
            cluster2.append(each)
        if what == 3:
            cluster3.append(each)
        if what == 4:
            print("something")

make_clusters(x)
print(cluster1)
print(cluster2)
print(cluster3)

[[5.9, 3.2], [4.6, 2.9], [4.7, 3.2], [5.0, 3.0], [4.9, 3.1], [5.1, 3.8], [6.0, 3.0]]
[[5.5, 4.2]]
[[6.2, 2.8], [6.7, 3.1]]


In [66]:
def get_centroid(cluster):
    x_sum = 0
    y_sum = 0
    
    for each in cluster:
        x_sum += each[0]
        y_sum += each[1]
    centr = [round(x_sum/len(cluster),3), round(y_sum/len(cluster),3)]
    return centr

def flush_clusters():
    cluster1 = []
    cluster2 = []
    cluster3 = []

In [46]:
# Answer for centroid of red cluster after one interation
centroid[0] = get_centroid(cluster1)
print("Centroid of Cluster 1 on One iteration: ",centroid[0])

Centroid of Cluster 1 on One iteration:  [5.171, 3.171]


In [47]:
# green cluster centroid after two iterations
cluster2_copy = cluster2
cent = get_centroid(cluster2_copy)
centroid[1] = cent
flush_clusters()
make_clusters(x)
cent = get_centroid(cluster2)
print("Centroid of Cluster 2 on second iteration: ",cent)

Centroid of Cluster 2 on second iteration:  [5.367, 4.067]


In [48]:
for i in range(0,5):
    cent = get_centroid(cluster3)
    centroid[2] = cent
    print(f"Iteration {i+1}: {cent}")
    flush_clusters()
    make_clusters(x)

Iteration 1: [6.283, 3.0]
Iteration 2: [6.25, 3.01]
Iteration 3: [6.236, 3.014]
Iteration 4: [6.228, 3.017]
Iteration 5: [6.223, 3.018]


The centroid of third cluster after converging is [6.22, 3.01]

In [67]:
for i in range(0,500):
    if i == 0:
        make_clusters(x)
    centroid[0] = get_centroid(cluster1)
    centroid[1] = get_centroid(cluster2)
    centroid[2] = get_centroid(cluster3)
    
    print(f"{i+1} Iteration: ", centroid)
    flush_clusters()
    make_clusters(x)

1 Iteration:  [[5.171, 3.171], [5.5, 4.2], [6.45, 2.95]]
2 Iteration:  [[5.089, 3.144], [5.4, 4.1], [6.325, 2.988]]
3 Iteration:  [[5.036, 3.127], [5.367, 4.067], [6.283, 3.0]]
4 Iteration:  [[5.0, 3.115], [5.35, 4.05], [6.263, 3.006]]
5 Iteration:  [[4.973, 3.107], [5.34, 4.04], [6.25, 3.01]]
6 Iteration:  [[4.953, 3.1], [5.333, 4.033], [6.242, 3.012]]
7 Iteration:  [[4.937, 3.095], [5.329, 4.029], [6.236, 3.014]]
8 Iteration:  [[4.924, 3.09], [5.325, 4.025], [6.231, 3.016]]
9 Iteration:  [[4.913, 3.087], [5.322, 4.022], [6.228, 3.017]]
10 Iteration:  [[4.904, 3.084], [5.32, 4.02], [6.225, 3.017]]
11 Iteration:  [[4.896, 3.081], [5.318, 4.018], [6.223, 3.018]]
12 Iteration:  [[4.89, 3.079], [5.317, 4.017], [6.221, 3.019]]
13 Iteration:  [[4.884, 3.077], [5.315, 4.015], [6.219, 3.019]]
14 Iteration:  [[4.879, 3.076], [5.314, 4.014], [6.218, 3.02]]
15 Iteration:  [[4.874, 3.074], [5.313, 4.013], [6.217, 3.02]]
16 Iteration:  [[4.87, 3.073], [5.312, 4.013], [6.216, 3.02]]
17 Iteration:  

The clusters take 400 iterations to converge